packages/medusa/src/services/product-variant-inventory.ts

```diff
@@ -598,6 +609,66 @@ class ProductVariantInventoryService extends TransactionBaseService {
       )
     }
   }
+
+  async setVariantAvailability(
+    variants: ProductVariant[] | PricedVariant[],
+    salesChannelId: string | undefined
+  ): Promise<ProductVariant[] | PricedVariant[]> {
+    if (!this.inventoryService_) {
+      return variants
+    }
+
+    return await Promise.all(
+      variants.map(async (variant) => {
+        if (!variant.id) {
+          return variant
+        }
+
+        if (!salesChannelId) {
+          delete variant.inventory_quantity
+          return variant
+        }
+
+        const variantInventory = await this.listByVariant(variant.id)
+
+        variant.inventory_quantity = Math.min(
+          ...(await Promise.all(
+            variantInventory.map(async (variantInventory) => {
+              return (
+                // eslint-disable-next-line max-len
+                (await this.salesChannelInventoryService_.retrieveAvailableItemQuantity(
+                  salesChannelId,
+                  variantInventory.inventory_item_id
+                )) / variantInventory.required_quantity
+              )
+            })
+          ))
+        )
```

Comment:
question: I don't completely follow what is happening here?

Reply:
We calculate the available quantity for the variant in the sales channel across inventoryItems weighted by the required quantity. 
i.e. we've previously gotten the `inventoryItem`'s based on the variant id (on line 632). 

We map that to retrieve the available quantity from the inventoryItemService (using the `salesChannelInventoryService` to fetch locations: 

```
async retrieveAvailableItemQuantity(
    salesChannelId: string,
    itemId: string
  ): Promise<number> {
    const locations = await this.salesChannelLocationService_.listLocations(
      salesChannelId
    )

    return await this.inventoryService_.retrieveAvailableQuantity(
      itemId,
      locations
    )
  }
```

We then adjust for the composition of the productVariant and how many inventoryItems are required to fulfil a single variant by diving with the `requiredQuantity`. 

This gives a list of all inventoryItems adjusted for how many we can fulfil and we return the smallest element of that list, since even though we have 12 table legs, if we only have 2 tabletops we can only fulfil 2 tables.

essentially this: https://github.com/medusajs/medusa/pull/2970#discussion_r1082270178

(though the get-inventory endpoint should be adjusted the same way)